src/topologies/halfedge.jl

```diff
@@ -332,45 +331,87 @@ Base.convert(::Type{HalfEdgeTopology}, t::Topology) = HalfEdgeTopology(collect(e
 # HELPER FUNCTIONS
 # -----------------
 
-function adjsort(elems::AbstractVector{<:Connectivity})
-  # initialize list of adjacent elements
-  # with first element from original list
-  list = indices.(elems)
-  adjs = Tuple[popfirst!(list)]
-
-  # the loop will terminate if the mesh
-  # is manifold, and that is always true
-  # with half-edge topology
-  while !isempty(list)
-    # lookup all elements that share at least
-    # one vertex with the last adjacent element
-    found = false
-    vinds = last(adjs)
-    for i in vinds
-      einds = findall(e -> i ∈ e, list)
-      if !isempty(einds)
-        # lookup all elements that share at
-        # least two vertices (i.e. edge)
-        for j in sort(einds, rev=true)
-          if length(vinds ∩ list[j]) > 1
-            found = true
-            push!(adjs, popat!(list, j))
-          end
-        end
+# permutation of elements in adjacent-first order
+function adjsortperm(elems::AbstractVector{<:Connectivity})
+  reduce(vcat, conneccomps(elems))
+end
+
+# connected components from list of elements
+function conneccomps(elems::AbstractVector{<:Connectivity})
+  # initialize list of connected components
+  comps = [[firstindex(elems)]]
+
+  # initialize list of seen vertices
+  seen = Set{Int}()
+  for v in indices(first(elems))
+    push!(seen, v)
+  end
+
+  # remaining elements to process
+  remaining = collect(eachindex(elems)[2:end])
+
+  added = false
+  while !isempty(remaining)
+    iter = 1
+    while iter ≤ length(remaining)
+      elem = elems[remaining[iter]]
+
+      # manually union-split two most common polytopes
+      # for type stability and maximum performance
+      isadjacent = if elem isa Connectivity{Triangle,3}
+        adjelem!(seen, elem)
+      elseif elem isa Connectivity{Quadrangle,4}
```

Comment:
```suggestion
      isadjacent = if elem isa Connectivity{Triangle}
        adjelem!(seen, elem)
      elseif elem isa Connectivity{Quadrangle}
```

Can we omit the second type parameter?

Reply:
Benchmarks indicate they are very much necessary. I expect it is related to the fact that `Connectivity{Triangle}` is a `UnionAll` type (incompletely specified).